### PR TITLE
dm/openpai: add more task api

### PR DESCRIPTION
**Pull request by Ehco1996**
### What problem does this PR solve? 
this is port of https://github.com/pingcap/dm/pull/2222

### What is changed and how it works?
as title, add more api about task

The api for the list table/schema is implemented by passing a json string using the Msg field in the rpc resp of the dm-worker.

close https://github.com/pingcap/dm/pull/2222

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Release note 

```release-note
`None`.
```

### CODE REVIEW

**Comment by Ehco1996** [comment]
@lichunzhu  PTAL😬

File: dm/dm/master/openapi.go [hunk 1]
```diff
@@ -557,43 +563,183 @@ func (s *Server) DMAPIGetTaskStatus(ctx echo.Context, taskName string, params op
...
+	newCtx := ctx.Request().Context()
+	resp, err := worker.SendRequest(newCtx, &workerReq, s.cfg.RPCTimeout)
+	if err != nil {
+		return err
+	}
+	if !resp.OperateSchema.Result {
+		return terror.ErrOpenAPICommonError.New(resp.OperateSchema.Msg)
+	}
+	tableList := openapi.TableNameList{}
+	if err := json.Unmarshal([]byte(resp.OperateSchema.Msg), &tableList); err != nil {
```
Comment (lichunzhu) [inline dm/dm/master/openapi.go:643]: ditto

File: dm/pkg/schema/tracker.go [hunk 2]
```diff
@@ -221,6 +221,21 @@ func (tr *Tracker) AllSchemas() []*model.DBInfo {
...
+	for _, db := range allSchemas {
+		if db.Name.String() == schema {
+			tables := make([]string, len(db.Tables))
+			for i, t := range db.Tables {
+				tables[i] = t.Name.String()
+			}
+			return tables, nil
+		}
+	}
+	return nil, errors.New("can not this schema in tracker")
```
Comment (lichunzhu) [inline dm/pkg/schema/tracker.go:236]: Do we need a terror?
```suggestion
	return nil, errors.New("can not find this schema in tracker")
```

File: dm/syncer/schema.go [hunk 3]
```diff
@@ -32,12 +33,33 @@ import (
...
 		Name:   req.Table,
 	}
 	switch req.Op {
+	case pb.SchemaOp_ListSchema:
+		allSchema := s.schemaTracker.AllSchemas()
+		schemaList := make([]string, len(allSchema))
+		for i, schema := range allSchema {
+			schemaList[i] = schema.Name.String()
+		}
+		schemaListJSON, err := json.Marshal(schemaList)
```
Comment (lichunzhu) [inline dm/syncer/schema.go:48]: terror

File: dm/syncer/schema.go [hunk 4]
```diff
@@ -32,12 +33,33 @@ import (
...
+		if err != nil {
+			return "", err
+		}
+		return string(schemaListJSON), err
+	case pb.SchemaOp_ListTable:
+		tables, err := s.schemaTracker.ListSchemaTables(req.Database)
+		if err != nil {
+			return "", err
+		}
+		tableListJSON, err := json.Marshal(tables)
```
Comment (lichunzhu) [inline dm/syncer/schema.go:58]: terror

File: dm/tests/openapi/run.sh [hunk 5]
```diff
@@ -269,6 +271,15 @@ function test_noshard_task() {
 	# get task list
 	openapi_task_check "get_task_list" 1
 
+	# stop  task
```
Comment (lichunzhu) [inline dm/tests/openapi/run.sh:274]: whatever
```suggestion
	# stop task
```

File: dm/tests/openapi/client/openapi_task_check [hunk 6]
```diff
@@ -176,6 +176,72 @@ def get_task_list(task_count):
...
+        url=url,
+        json=[
+            source_name,
+        ],
+    )
+    assert resp.status_code == 200
+
+
+def operate_schema_and_table_success(task_name, source_name):
+    schema_url = API_ENDPOINT + "/" + task_name + "/sources/" + source_name + "/schemas"
```
Comment (lichunzhu) [inline dm/tests/openapi/client/openapi_task_check:202]: Just asking, do we have a test for `task_ name/source_name` with special character?

File: dm/tests/openapi/client/openapi_task_check [hunk 7]
```diff
@@ -176,6 +176,72 @@ def get_task_list(task_count):
...
+    resp = requests.get(url=schema_url)
+    assert resp.status_code == 200
+    print("get_task_schema_success schema resp=", resp.json())
+    assert len(resp.json()) > 0
```
Comment (lichunzhu) [inline dm/tests/openapi/client/openapi_task_check:206]: Do we need to check the content here?

File: dm/dm/master/openapi.go [hunk 8]
```diff
@@ -479,9 +481,13 @@ func (s *Server) DMAPIGetTaskList(ctx echo.Context) error {
 
 // DMAPIGetTaskStatus url is:(GET /api/v1/tasks/{task-name}/status).
```
[the remark this answers is not in the document]
Comment (Ehco1996) [reply dm/dm/master/openapi.go:482]: no, you will got a 301 redirect 
```bash
➜ curl -v http://127.0.0.1:8261/api/v1/tasks//status
*   Trying 127.0.0.1:8261...
* Connected to 127.0.0.1 (127.0.0.1) port 8261 (#0)
> GET /api/v1/tasks//status HTTP/1.1
> Host: 127.0.0.1:8261
> User-Agent: curl/7.77.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 301 Moved Permanently
< Access-Control-Allow-Headers: accept, content-type, authorization
< Access-Control-Allow-Methods: POST, GET, OPTIONS, PUT, DELETE
< Access-Control-Allow-Origin: *
< Content-Type: text/html; charset=utf-8
< Location: /api/v1/tasks/status
< Date: Mon, 01 Nov 2021 06:33:32 GMT
< Content-Length: 55
<
<a href="/api/v1/tasks/status">Moved Permanently</a>.

* Connection #0 to host 127.0.0.1 left intact
```

File: dm/tests/openapi/client/openapi_task_check (hunk 6 above)
Comment (Ehco1996) [reply dm/tests/openapi/client/openapi_task_check:203]: yes, becasuse JSON allows for both escaped or non-escaped non-ascii characters.

File: dm/tests/openapi/client/openapi_task_check (hunk 7 above)
Comment (Ehco1996) [reply dm/tests/openapi/client/openapi_task_check]: add more check in https://github.com/pingcap/ticdc/pull/3171/commits/b4575a1b56c20ad7553fd61779c6534a4398dea2

File: dm/pkg/schema/tracker.go [hunk 9]
```diff
@@ -221,6 +221,21 @@ func (tr *Tracker) AllSchemas() []*model.DBInfo {
 	return filteredSchemas
 }
 
+// ListSchemaTables lists all tables in the schema.
+func (tr *Tracker) ListSchemaTables(schema string) ([]string, error) {
+	allSchemas := tr.AllSchemas()
+	for _, db := range allSchemas {
+		if db.Name.String() == schema {
```
Comment (glorv) [inline dm/pkg/schema/tracker.go:229]: Do we need to make this check case-sensitive? 🤔️

File: dm/dm/master/openapi.go [hunk 10]
```diff
@@ -557,43 +563,183 @@ func (s *Server) DMAPIGetTaskStatus(ctx echo.Context, taskName string, params op
...
+	newCtx := ctx.Request().Context()
+	resp, err := worker.SendRequest(newCtx, &workerReq, s.cfg.RPCTimeout)
+	if err != nil {
+		return err
+	}
+	if !resp.OperateSchema.Result {
+		return terror.ErrOpenAPICommonError.New(resp.OperateSchema.Msg)
+	}
+	schemaList := openapi.SchemaNameList{}
+	if err := json.Unmarshal([]byte(resp.OperateSchema.Msg), &schemaList); err != nil {
```
[the remark this answers is not in the document]
Comment (Ehco1996) [reply dm/dm/master/openapi.go:613]: add terror in https://github.com/pingcap/ticdc/pull/3171/commits/3a1e1eda14dd03a69e7fa11213bd1809927e59e0

**Review by lichunzhu — APPROVED** [review]

File: dm/dm/master/openapi.go [hunk 11]
```diff
@@ -557,43 +563,183 @@ func (s *Server) DMAPIGetTaskStatus(ctx echo.Context, taskName string, params op
...
+		return terror.ErrSchemaTrackerUnMarshalJSON.Delegate(err, resp.OperateSchema.Msg)
+	}
+	return ctx.JSON(http.StatusOK, schemaList)
 }
 
-// DMAPIGetTaskSourceTableList get task source table list url is: (GET /api/v1/tasks/{task-name}/sources/{source-name}/schemas/{schema-name}).
-func (s *Server) DMAPIGetTaskSourceTableList(ctx echo.Context, taskName string, sourceName string, schemaName string) error {
-	return nil
+// DMAPIGetTableListByTaskAndSource get task source table list url is: (GET /api/v1/tasks/{task-name}/sources/{source-name}/schemas/{schema-name}).
+func (s *Server) DMAPIGetTableListByTaskAndSource(ctx echo.Context, taskName string, sourceName string, schemaName string) error {
```
Comment (lance6716) [inline dm/dm/master/openapi.go:620]: I remember this API is to check the tables in allow list before we run a task? At that time the schema tracker doesn't contain any data

**Comment by lance6716** [comment]
/hold

waiting for comments

File: dm/dm/master/openapi.go (hunk 11 above)
Comment (Ehco1996) [reply dm/dm/master/openapi.go:620]: this api is used to get table list in trackcer (maybe some prepare work for operate-scema)

the api to get >the tables in allow list before we run a task is in [here](https://github.com/pingcap/ticdc/blob/dc32dd69c4c1384acff27b4631d25eab38f27697/dm/dm/master/openapi.go#L283)

File: dm/pkg/schema/tracker.go (hunk 9 above)
Comment (Ehco1996) [reply dm/pkg/schema/tracker.go:229]: The current impl is case-sensitive 😬

**Comment by Ehco1996** [comment]
> /hold
> 
> waiting for comments to be addressed

all comments are addressed, do you have more comments? @lance6716

**Pull request merged**